File: dotnet/test/Azure.Iot.Operations.Protocol.MetlTests/TestCaseExecutor.cs
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

namespace Azure.Iot.Operations.Protocol.MetlTests
{
    public class TestCaseExecutor
    {
        public static string? DefaultCommandName;
        public static TestCaseSerializer DefaultSerializer = new();
        public static string? DefaultRequestTopic;
        public static string? DefaultExecutorId;
        public static string? DefaultTopicNamespace;
        public static bool DefaultIdempotent;
        public static TestCaseDuration? DefaultCacheTtl;
        public static TestCaseDuration? DefaultExecutorTimeout;
        public static Dictionary<string, string[]>? DefaultRequestResponsesMap;
        public static int? DefaultExecutionConcurrency;

        public string? CommandName { get; set; } = DefaultCommandName;

        public TestCaseSerializer Serializer { get; set; } = DefaultSerializer;

        public string? RequestTopic { get; set; } = DefaultRequestTopic;

        public string? ExecutorId { get; set; } = DefaultExecutorId;

        public string? TopicNamespace { get; set; } = DefaultTopicNamespace;

        public Dictionary<string, string>? TopicTokenMap { get; set; }

        public bool Idempotent { get; set; } = DefaultIdempotent;

        public TestCaseDuration? CacheTtl { get; set; } = DefaultCacheTtl;

        public TestCaseDuration? ExecutionTimeout { get; set; } = DefaultExecutorTimeout;

        public Dictionary<string, string[]> RequestResponsesMap { get; set; } = DefaultRequestResponsesMap ?? new();

        public Dictionary<string, string?> ResponseMetadata { get; set; } = new();

        public string? TokenMetadataPrefix { get; set; }

        public int? ExecutionConcurrency { get; set; } = DefaultExecutionConcurrency;

        public bool RaiseError { get; set; }

        public List<TestCaseSync> Sync { get; set; } = new();
    }
}

File: go/test/protocol/test_case_executor.go
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.
package protocol

import (
	"gopkg.in/yaml.v3"
)

type testCaseExecutor struct {
	CommandName          *string             `yaml:"command-name"`
	RequestTopic         *string             `yaml:"request-topic"`
	ExecutorID           *string             `yaml:"executor-id"`
	TopicNamespace       *string             `yaml:"topic-namespace"`
	TopicTokenMap        map[string]string   `yaml:"topic-token-map"`
	Idempotent           bool                `yaml:"idempotent"`
	ExecutionTimeout     *TestCaseDuration   `yaml:"execution-timeout"`
	RequestResponsesMap  map[string][]string `yaml:"request-responses-map"`
	ResponseMetadata     map[string]*string  `yaml:"response-metadata"`
	ExecutionConcurrency *uint               `yaml:"execution-concurrency"`
	RaiseError           TestCaseError       `yaml:"raise-error"`
	Sync                 []TestCaseSync      `yaml:"sync"`
}

type TestCaseExecutor struct {
	testCaseExecutor
}

func (executor *TestCaseExecutor) UnmarshalYAML(node *yaml.Node) error {
	*executor = TestCaseExecutor{}

	executor.CommandName = TestCaseDefaultInfo.Prologue.Executor.GetCommandName()
	executor.RequestTopic = TestCaseDefaultInfo.Prologue.Executor.GetRequestTopic()
	executor.ExecutorID = TestCaseDefaultInfo.Prologue.Executor.GetExecutorID()
	executor.TopicNamespace = TestCaseDefaultInfo.Prologue.Executor.GetTopicNamespace()
	executor.Idempotent = TestCaseDefaultInfo.Prologue.Executor.GetIdempotent()
	executor.RequestResponsesMap = TestCaseDefaultInfo.Prologue.Executor.GetRequestResponsesMap()
	executor.ExecutionConcurrency = TestCaseDefaultInfo.Prologue.Executor.GetExecutionConcurrency()

	err := node.Decode(&executor.testCaseExecutor)

	if executor.ExecutionTimeout == nil {
		defaultExecutionTimeout := TestCaseDefaultInfo.Prologue.Executor.GetExecutionTimeout()
		executor.ExecutionTimeout = &defaultExecutionTimeout
	}

	return err
}
